Split EditorData class into 5 classes for better single responsibility.

EricEzaM opened the pull request:
This has no external scripting API changes. This change affects internal usage only. This better encapsulates individual responsibilities for EditorData. Previously EditorData was a big set of, for the most part, unrelated methods and data, making it quite confusing to know exactly what it was for.

Note all changes are in `editor_data.h`, `editor_data.cpp` and `editor_node.h`. All the rest of the changes are just replacing `EditorNode::get_editor_data()` calls with the new equivalent new methods in EditorNode.

This splits up EditorData into 5 different classes.
They are accessed via EditorNode as EditorData previously was (e.g. `EditorNode::get_editor_clipboard()`).

EditorData becomes:
1. EditorClipboard - manages the shared clipboard of the editor (e.g. when copying and pasting object parameters)
2. EditorUndoRedo - the storage location for the one central UndoRedo class for the Editor. This one is optional TBH. The UndoRedo could just be made directly in EditorNode, but I kept it as a separate class for now, as I thought maybe direct storage of state in EditorNode is to be avoided? If that is not the case, EditorUndoRedo could be removed and just instance the UndoRedo in EditorNode.
3. EditorPluginsManager - manages EditorPlugins.
4. EditorScenesManager - manages scenes which are currently open.
5. EditorCustomTypes - manages registration of custom node types from scripts and editor plugins.

** The "manager" names are new to Godot it seems... I am definitely open to alternative names. Just note that "EditorPlugins" is already taken. If anything, I propose changing "EditorPlugins" to "ExternalEditorPlugins" (as that is what that class is used for), and then renaming these "managers" to "EditorPlugins" and "EditorScenes" respectively.

EricEzaM commented:
I still think this is a valuable addition in order to clean up the codebase. I won't rebase until I get some further feedback and support from other devs.

I know reduz is looking to change the way undo/redo works, such that there isn't just one central undo/redo: https://github.com/godotengine/godot-proposals/issues/2109 so this might be affected by his plans for that.

groud reviewed (approved):
I believe the change makes sense. Splitting things into smaller pieces makes sense to me.
It needs a rebase though.

EricEzaM commented:
Rebased.

JFonS commented:
I'm a bit wary of the "Manager" at the end of class names. I feel like it doesn't really mean anything. I'd prefer going with the renames you mention in the original PR post.

That being said, I don't feel strongly about it, just personal preference.

EricEzaM commented:
@JFonS I agree, they are at the very least a point of contention. I will look into renaming EditorPlugins to ExternalEditorPlugins and then removing the `Manager` suffixes.

EricEzaM committed:
Rename EditorPlugins to CustomEditorPlugins

EricEzaM commented:
Ok, EditorPlugins -> CustomEditorPlugins. `Manager` suffixes removed.

EricEzaM commented:
@akien-mga Please the static check error. It was me to increase the tabs for some reason. However, when I do that, the pre-commit hook wants me to revert those tabs...

akien-mga commented:
@EricEzaM You should update to clang-format 11 as used in CI.

EricEzaM committed:
Split EditorData class into 5 classes for better single responsibility.

This splits up EditorData into 5 different classes.
They are accessed via EditorNode as EditorData previously was, except now their get_* functions are static, so accessing them is much shorter:
Before: EditorNode::get_singleton()->get_editor_data()
Now: EditorNode::get_editor_clipboard()

EditorData becomes:
1. EditorClipboard - manages the shared clipboard of the editor (e.g. when copying and pasting object parameters)
2. EditorUndoRedo - the storage location for the one central UndoRedo class for the Editor.
3. EditorPluginsManager - manages EditorPlugins.
4. EditorScenesManager - manages scenes which are currently open.
5. EditorCustomTypes - manages registration of custom node types from scripts and editor plugins.

EricEzaM commented:
@akien-mga Hopefully that has done it - thanks!

EricEzaM commented:
Waiting for Godot 4.0....

Waiting before making such changes, as per contibutors meeting 4/06/2021. This can be revisited after 4.0 and the core API has stabilised.
